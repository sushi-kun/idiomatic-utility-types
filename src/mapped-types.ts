/**
 * Omit (complements Pick)
 * @desc From `T` remove a set of properties `K`
 * @example
 *   type Props = { name: string; age: number; visible: boolean };
 *
 *   // Expect: { name: string; visible: boolean; }
 *   type RequiredProps = Omit<Props, 'age'>;
 */
export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
