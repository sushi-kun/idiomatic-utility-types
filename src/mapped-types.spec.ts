import { testType } from '../utils/test-utils';
import { Omit } from './mapped-types';

/**
 * Fixtures
 */

type Props = { name: string; age: number; visible: boolean };
type DefaultProps = { age: number };
type NewProps = { age: string; other: string };
type MixedProps = { name: string; setName: (name: string) => void };
type ReadWriteProps = { readonly a: number; b: string };

// @dts-jest:group Omit
describe('mapped types', () => {
  it('Omit', () => {
    // @dts-jest:pass:snap
    testType<Omit<Props, 'age'>>();
    // @dts-jest:pass:snap
    testType<Omit<Props | NewProps, 'age'>>();
  });
});
