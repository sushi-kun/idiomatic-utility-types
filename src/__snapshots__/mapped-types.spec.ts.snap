// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Omit testType<Omit<Props | NewProps, 'age'>>() (type) should match snapshot: Omit testType<Omit<Props | NewProps, 'age'>>() 1`] = `"Pick<Props | NewProps, never>"`;

exports[`Omit testType<Omit<Props, 'age'>>() (type) should match snapshot: Omit testType<Omit<Props, 'age'>>() 1`] = `"Pick<Props, \\"name\\" | \\"visible\\">"`;
